services/codacy/codacy-coverage.service.js

```diff
@@ -1,78 +1,82 @@
+      base: 'codacy/coverage',
+      format: '(?!grade/)([^/]+)(?:/(.+))?',
+      capture: ['projectId', 'branch'],
     }
   }
 
   static get examples() {
     return [
       {
         title: 'Codacy coverage',
-        previewUrl: 'coverage/59d607d0e311408885e418004068ea58',
+        urlPattern: ':projectId',
+        staticExample: this.render({ percentage: 90 }),
+        exampleUrl: '59d607d0e311408885e418004068ea58',
       },
       {
         title: 'Codacy branch coverage',
-        previewUrl: 'coverage/59d607d0e311408885e418004068ea58/master',
+        urlPattern: ':projectId/:branch',
+        staticExample: this.render({ percentage: 90 }),
+        exampleUrl: '59d607d0e311408885e418004068ea58/master',
       },
     ]
   }
 
-  static registerLegacyRouteHandler({ camp, cache }) {
-    camp.route(
-      /^\/codacy\/coverage\/(?!grade\/)([^/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
-      cache((data, match, sendBadge, request) => {
-        const projectId = match[1] // eg. e27821fb6289410b8f58338c7e0bc686
-        const branch = match[2]
-        const format = match[3]
+  static get defaultBadgeData() {
+    return {
+      label: 'coverage',
+    }
+  }
 
-        const queryParams = {}
-        if (branch) {
-          queryParams.branch = branch
-        }
-        const query = queryString.stringify(queryParams)
-        const url = `https://api.codacy.com/project/badge/coverage/${projectId}?${query}`
-        const badgeData = getBadgeData('coverage', data)
-        fetchFromSvg(
-          request,
-          url,
-          /text-anchor="middle">([^<>]+)<\/text>/,
-          (err, res) => {
-            if (err != null) {
-              badgeData.text[1] = 'inaccessible'
-              sendBadge(format, badgeData)
-              return
-            }
-            try {
-              const coverage = parseInt(res)
-              if (Number.isNaN(coverage)) {
-                badgeData.text[1] = 'unknown'
-              } else {
-                badgeData.text[1] = res
-                badgeData.colorscheme = coveragePercentageColor(coverage)
-              }
-              sendBadge(format, badgeData)
-            } catch (e) {
-              badgeData.text[1] = 'invalid'
-              sendBadge(format, badgeData)
-            }
-          }
-        )
+  static render({ percentage }) {
+    return {
+      message: `${percentage}%`,
+      color: coveragePercentageColor(percentage),
+    }
+  }
+
+  static transform({ coverageString }) {
+    return {
+      percentage: parseFloat(coverageString.replace(/%$/, '')),
+    }
+  }
+
+  async handle({ projectId, branch }) {
+    const { message: coverageString } = await this._requestSvg({
+      schema,
+      url: `https://api.codacy.com/project/badge/coverage/${encodeURIComponent(
+        projectId
+      )}`,
+      options: { qs: { branch } },
+      valueMatcher: /text-anchor="middle">([^<>]+)<\/text>/,
+    })
+    if (coverageString === '!') {
```

Comment:
I don't think "not found" is the correct meaning of the exclamation mark. Consider project `e27821fb6289410b8f58338c7e0bc686`:
- coverage: ![](https://api.codacy.com/project/badge/coverage/e27821fb6289410b8f58338c7e0bc686) (https://api.codacy.com/project/badge/coverage/e27821fb6289410b8f58338c7e0bc686)
- grade: ![](https://api.codacy.com/project/badge/grade/e27821fb6289410b8f58338c7e0bc686) (https://api.codacy.com/project/badge/grade/e27821fb6289410b8f58338c7e0bc686)

The project is found and as far as I can tell, `!` is equivalent to "coverage unknown" or "coverage  not available for this project". When a project is not found a 404 is returned.

Reply:
Thanks for digging into that. I'll see if I can't get a separate "project not found" test case working here.